File: ls/server/src/main/java/org/opencds/cqf/cql/ls/server/utility/Diagnostics.java
package org.opencds.cqf.cql.ls.server.utility;

import java.util.HashSet;
import java.util.Set;
import org.cqframework.cql.cql2elm.CqlCompilerException;
import org.eclipse.lsp4j.Diagnostic;
import org.eclipse.lsp4j.DiagnosticSeverity;
import org.eclipse.lsp4j.Position;
import org.eclipse.lsp4j.Range;

public class Diagnostics {
    private Diagnostics() {}

    public static Diagnostic convert(CqlCompilerException error) {
        if (error.getLocator() != null) {
            Range range = position(error);
            Diagnostic diagnostic = new Diagnostic();
            DiagnosticSeverity severity = severity(error.getSeverity());

            diagnostic.setSeverity(severity);
            diagnostic.setRange(range);
            diagnostic.setMessage(error.getMessage());

            return diagnostic;
        } else {
            return null;
        }
    }

    public static Set<Diagnostic> convert(Iterable<CqlCompilerException> errors) {
        Set<Diagnostic> result = new HashSet<>();
        for (CqlCompilerException error : errors) {
            Diagnostic diagnostic = convert(error);
            if (diagnostic != null) {
                result.add(diagnostic);
            }
        }
        return result;
    }

    private static DiagnosticSeverity severity(CqlCompilerException.ErrorSeverity severity) {
        switch (severity) {
            case Error:
                return DiagnosticSeverity.Error;
            case Warning:
                return DiagnosticSeverity.Warning;
            case Info:
            default:
                return DiagnosticSeverity.Information;
        }
    }

    private static Range position(CqlCompilerException error) {
        // The Language server API assumes 0 based indices and an exclusive range
        return new Range(
                new Position(
                        Math.max(error.getLocator().getStartLine() - 1, 0),
                        Math.max(error.getLocator().getStartChar() - 1, 0)),
                new Position(
                        Math.max(error.getLocator().getEndLine() - 1, 0),
                        Math.max(error.getLocator().getEndChar(), 0)));
    }
}

File: ls/server/src/main/kotlin/org/opencds/cqf/cql/ls/server/utility/Diagnostics.kt
package org.opencds.cqf.cql.ls.server.utility

import org.cqframework.cql.cql2elm.CqlCompilerException
import org.cqframework.cql.cql2elm.tracking.TrackBack
import org.eclipse.lsp4j.Diagnostic
import org.eclipse.lsp4j.DiagnosticSeverity
import org.eclipse.lsp4j.Position
import org.eclipse.lsp4j.Range

object Diagnostics {
    @JvmStatic
    fun convert(error: CqlCompilerException): Diagnostic? {
        val locator = error.locator ?: return null
        val range = position(locator)
        val diagnostic = Diagnostic()
        diagnostic.severity = severity(error.severity)
        diagnostic.range = range
        diagnostic.message = error.message
        return diagnostic
    }

    @JvmStatic
    fun convert(errors: Iterable<CqlCompilerException>): Set<Diagnostic> = errors.mapNotNull { convert(it) }.toSet()

    private fun severity(severity: CqlCompilerException.ErrorSeverity): DiagnosticSeverity {
        return when (severity) {
            CqlCompilerException.ErrorSeverity.Error -> DiagnosticSeverity.Error
            CqlCompilerException.ErrorSeverity.Warning -> DiagnosticSeverity.Warning
            else -> DiagnosticSeverity.Information
        }
    }

    private fun position(locator: TrackBack): Range {
        return Range(
            Position(
                maxOf(locator.startLine - 1, 0),
                maxOf(locator.startChar - 1, 0),
            ),
            Position(
                maxOf(locator.endLine - 1, 0),
                maxOf(locator.endChar, 0),
            ),
        )
    }
}
